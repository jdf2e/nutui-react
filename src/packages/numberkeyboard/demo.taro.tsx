import React, { useState } from 'react'
import Taro from '@tarojs/taro'
import { useTranslate } from '@/sites/assets/locale/taro'
import { Cell, Toast, NumberKeyboard } from '@/packages/nutui.react.taro'
import Header from '@/sites/components/header'

const NumberKeyboardDemo = () => {
  const [translated] = useTranslate({
    'zh-CN': {
      basic: '默认键盘',
      sidebar: '带右侧栏键盘',
      confirmText: '支付',
      randomKeyOrder: '随机数键盘',
      title: '标题',
      withTitle: '带标题栏键盘',
      idNumberKeyboard: '身份证键盘',
      popup: '透传 Popup 属性',
    },
    'zh-TW': {
      basic: '默認鍵盤',
      sidebar: '帶右側欄鍵盤',
      confirmText: '支付',
      randomKeyOrder: '隨機數鍵盤',
      title: '標題',
      withTitle: '帶標題欄鍵盤',
      idNumberKeyboard: '身份證鍵盤',
      popup: '透傳 Popup 屬性',
    },
    'en-US': {
      basic: 'Default Keyboard',
      sidebar: 'Keyboard With Sidebar',
      confirmText: 'pay',
      randomKeyOrder: 'Random Key Order',
      title: 'title',
      withTitle: 'Show Keyboard With Title',
      idNumberKeyboard: 'Show IdNumber Keyboard',
      popup: 'Use Popup Props',
    },
  })
  const [visible1, setVisible1] = useState(false)
  const [visible2, setVisible2] = useState(false)
  const [visible3, setVisible3] = useState(false)
  const [visible4, setVisible4] = useState(false)
  const [visible5, setVisible5] = useState(false)
  const [visible6, setVisible6] = useState(false)
  const onChange = (number: string) => {
    // Toast.text(`输入：${number}`)
    toastShow(`输入：${number}`)
  }
  const onDelete = () => {
    // Toast.text('删除')
    toastShow('删除')
  }
  const [show, SetShow] = useState(false)
  const [toastMsg, SetToastMsg] = useState('')
  const toastShow = (msg: any) => {
    SetToastMsg(msg)
    SetShow(true)
  }
  return (
    <>
      <Header />
      <div className={`demo ${Taro.getEnv() === 'WEB' ? 'web' : ''}`}>
        <h2>{translated.basic}</h2>
        <Cell
          title={translated.basic}
          onClick={() => {
            setVisible1(true)
          }}
        />
        <NumberKeyboard
          visible={visible1}
          onChange={onChange}
          onDelete={onDelete}
          onClose={() => setVisible1(false)}
        />
        <h2>{translated.sidebar}</h2>
        <Cell
          title={translated.sidebar}
          onClick={() => {
            setVisible2(true)
          }}
        />
        <NumberKeyboard
          visible={visible2}
          type="rightColumn"
          custom={['.', 'x']}
          onChange={onChange}
          onDelete={onDelete}
          onClose={() => setVisible2(false)}
        />
        <h2>{translated.randomKeyOrder}</h2>
        <Cell
          title={translated.randomKeyOrder}
          onClick={() => {
            setVisible3(true)
          }}
        />
        <NumberKeyboard
          visible={visible3}
          random
          onChange={onChange}
          onDelete={onDelete}
          onClose={() => setVisible3(false)}
        />
        <h2>{translated.withTitle}</h2>
        <Cell
          title={translated.withTitle}
          onClick={() => {
            setVisible4(true)
          }}
        />
        <NumberKeyboard
          visible={visible4}
          title={translated.title}
          custom={['.']}
          onChange={onChange}
          onDelete={onDelete}
          onClose={() => setVisible4(false)}
        />
        <h2>{translated.idNumberKeyboard}</h2>
        <Cell
          title={translated.idNumberKeyboard}
          onClick={() => {
            setVisible5(true)
          }}
        />
        <NumberKeyboard
          visible={visible5}
          custom={['X']}
          onChange={onChange}
          onDelete={onDelete}
          onClose={() => setVisible5(false)}
        />
        <h2>{translated.popup}</h2>
        <Cell
          title={translated.popup}
          onClick={() => {
            setVisible6(true)
          }}
        />
        <NumberKeyboard
          visible={visible6}
          onChange={onChange}
          onDelete={onDelete}
          onClose={() => setVisible6(false)}
          duration={1}
          overlayClassName="number-keyboard-overlay"
          onOpen={() => {
            toastShow('onOpen')
          }}
        />
        <Toast
          type="text"
          visible={show}
          msg={toastMsg}
          onClose={() => {
            SetShow(false)
          }}
        />
      </div>
    </>
  )
}

export default NumberKeyboardDemo
