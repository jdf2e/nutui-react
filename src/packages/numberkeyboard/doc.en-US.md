#  NumberKeyboard 

### Introduce

Virtual numeric keypad, used for scenarios where payment passwords are entered.

### Install
```ts
// react
import { NumberKeyboard } from '@nutui/nutui-react';
```


## Demo

### Default Keyboard
:::demo
```tsx
import React, { useState } from "react";
import { Cell,NumberKeyboard,Toast } from '@nutui/nutui-react';

const App = () => {
  const [visible,setVisible] = useState(false)
  const onChange = (number: string) => {
    Toast.text(`enter:${number}`)
  }
  const onDelete = () => {
    Toast.text('delete')
  }
  return (
     <>
       <Cell
        title="Default Keyboard"
        isLink
        onClick={() => {
          setVisible(true)
        }}
       />
        <NumberKeyboard
        visible={visible}
        onChange={onChange}
        onDelete={onDelete}
        onClose={() => setVisible(false)}
      />
     </>
  )
}
export default App;

```
:::

### Keyboard With Sidebar
:::demo
```tsx
import React, { useState } from "react";
import { Cell,NumberKeyboard,Toast } from '@nutui/nutui-react';

const App = () => {
  const [visible,setVisible] = useState(false)
  const onChange = (number: string) => {
    Toast.text(`enter:${number}`)
  }
  const onDelete = () => {
    Toast.text('delete')
  }
  return (
     <>
       <Cell
        title="Keyboard With Sidebar"
        isLink
        onClick={() => {
          setVisible(true)
        }}
       />
        <NumberKeyboard
        visible={visible}
        type="rightColumn"
        customKey={['.', 'x']}
        onChange={onChange}
        onDelete={onDelete}
        onClose={() => setVisible(false)}
      />
     </>
  )
}
export default App;

```
:::

###  Random Key Order
:::demo
```tsx
import React, { useState } from "react";
import { Cell,NumberKeyboard,Toast } from '@nutui/nutui-react';

const App = () => {
  const [visible,setVisible] = useState(false)
  const onChange = (number: string) => {
    Toast.text(`enter:${number}`)
  }
  const onDelete = () => {
    Toast.text('delete')
  }
  return (
     <>
       <Cell
        title=" Random Key Order"
        isLink
        onClick={() => {
          setVisible(true)
        }}
       />
        <NumberKeyboard
        visible={visible}
        randomKeys
        onChange={onChange}
        onDelete={onDelete}
        onClose={() => setVisible(false)}
      />
     </>
  )
}
export default App;

```
:::
### Show Keyboard With Title
:::demo
```tsx
import React, { useState } from "react";
import { Cell,NumberKeyboard,Toast } from '@nutui/nutui-react';

const App = () => {
  const [visible,setVisible] = useState(false)
  const onChange = (number: string) => {
    Toast.text(`enter:${number}`)
  }
  const onDelete = () => {
    Toast.text('delete')
  }
  return (
     <>
       <Cell
        title="Show Keyboard With Title"
        isLink
        onClick={() => {
          setVisible(true)
        }}
       />
        <NumberKeyboard
        visible={visible}
        title="title"
        customKey={['.']}
        onChange={onChange}
        onDelete={onDelete}
        onClose={() => setVisible(false)}
      />
     </>
  )
}
export default App;

```
:::
### Show IdNumber Keyboard
:::demo
```tsx
import React, { useState } from "react";
import { Cell,NumberKeyboard,Toast } from '@nutui/nutui-react';

const App = () => {
  const [visible,setVisible] = useState(false)
  const onChange = (number: string) => {
    Toast.text(`enter:${number}`)
  }
  const onDelete = () => {
    Toast.text('delete')
  }
  return (
     <>
       <Cell
        title="Show IdNumber Keyboard"
        isLink
        onClick={() => {
          setVisible(true)
        }}
       />
        <NumberKeyboard
        visible={visible}
        customKey={['X']}
        onChange={onChange}
        onDelete={onDelete}
        onClose={() => setVisible(false)}
      />
     </>
  )
}
export default App;

```
:::



## API

### Props

| Attribute | Description | Type | Default |
|----- | ----- | ----- | ----- |
| visible | Whether to show keyboard  | boolean | `false` | 
| title | Keyboard title | string | - |
| type | Keyboard type  | string | `default`, `rightColumn` |
| randomKeys | Whether to shuffle the order of keys  | boolean | `false` |
| customKey | Content of bottom left key   | string[] |  Array form supports adding up to two |
| confirmText  | Custom done button text,Such as "pay", "next", "submit" | string | `done` |
| pop-class    | Custom bullet box classname     | string         | -             |


### Event

| Event | Description | Arguments
|----- | ----- | -----
| onChange  | Emitted when a key is pressed                   | `value: string` |
| onDelete | 	Emitted when the delete key is pressed                | -             |
| onClose  | Emitted when the close button or non-keyboard area is clicked is clicked   | -             |


## Theming

### CSS Variables

The component provides the following CSS variables, which can be used to customize styles. Please refer to [ConfigProvider component](#/en-US/component/configprovider).

| Name | Default Value |
| --- | --- |
| --nutui-numberkeyboard-padding | `0 0 22px 0` |
| --nutui-numberkeyboard-background-color | `#f2f3f5` |
| --nutui-numberkeyboard-header-height | `34px` |
| --nutui-numberkeyboard-header-padding | `6px 0 0 0` |
| --nutui-numberkeyboard-header-color | `#646566` |
| --nutui-numberkeyboard-header-font-size | `16px` |
| --nutui-numberkeyboard-header-close-padding | `0 16px` |
| --nutui-numberkeyboard-header-close-color | `#576b95` |
| --nutui-numberkeyboard-header-close-font-size | `14px` |
| --nutui-numberkeyboard-header-close-background-color | `transparent` |
| --nutui-numberkeyboard-key-background-color | `#fff` |
| --nutui-numberkeyboard-key-active-background-color | `#ebedf0` |
| --nutui-numberkeyboard-key-height | `48px` |
| --nutui-numberkeyboard-key-line-height | `1.5` |
| --nutui-numberkeyboard-key-border-radius | `8px` |
| --nutui-numberkeyboard-key-font-size | `28px` |
| --nutui-numberkeyboard-key-font-color | `#333` |
| --nutui-numberkeyboard-key-finish-font-size | `16px` |
| --nutui-numberkeyboard-key-finish-font-color | `#fff` |
| --nutui-numberkeyboard-key-finish-background-color | `#1989fa` |
