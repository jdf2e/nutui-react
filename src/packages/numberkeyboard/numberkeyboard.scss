@import '../popup/popup.scss';

.nut-numberkeyboard {
  width: $numberkeyboard-width;
  padding: $numberkeyboard-padding;
  background-color: $numberkeyboard-background-color;
  user-select: none;
  background-color: $gray4;

  .nut-numberkeyboard__header {
    position: relative;
    display: flex;
    align-items: center;
    justify-content: center;
    box-sizing: content-box;
    height: $numberkeyboard-header-height;
    padding: $numberkeyboard-header-padding;
    color: $numberkeyboard-header-color;
    font-size: $numberkeyboard-header-font-size;

    .nut-numberkeyboard__header__tit {
      color: $title-color;
      display: inline-block;
    }

    .nut-numberkeyboard__header__close {
      position: absolute;
      display: block;
      right: 0;
      padding: $numberkeyboard-header-close-padding;
      color: $numberkeyboard-header-close-color;
      font-size: $numberkeyboard-header-close-font-size;
      background-color: $numberkeyboard-header-close-background-color;
      border: none;
      cursor: pointer;
    }
  }

  .nut-numberkeyboard__body {
    display: flex;
    padding: 6px 0 0 6px;

    .nut-numberkeyboard__body__keys {
      display: flex;
      flex: 3;
      flex-wrap: wrap;
    }

    .keyboard-wrapper {
      position: relative;
      flex: 1;
      flex-basis: 33%;
      box-sizing: border-box;
      padding: 0 6px 6px 0;
      background-color: $gray5;

      .key {
        display: flex;
        align-items: center;
        justify-content: center;
        height: $numberkeyboard-key-height;
        font-size: $numberkeyboard-key-font-size;
        line-height: $numberkeyboard-key-line-height;
        background-color: $numberkeyboard-key-background-color;
        color: $numberkeyboard-key-font-color;
        border-radius: $numberkeyboard-key-border-radius;
        cursor: pointer;
        &:active {
          background-color: $numberkeyboard-key-active-background-color;
        }
      }

      img {
        width: 30px;
        height: 24px;
      }
    }

    .keyboard-wrapper-large {
      flex-basis: 66%;
    }

    .nut-numberkeyboard__sidebar {
      display: flex;
      flex: 1;
      flex-direction: column;

      .keyboard-wrapper {
        .key {
          position: absolute;
          top: 0;
          right: 6px;
          bottom: 6px;
          left: 0;
          height: auto;
        }

        .finish {
          font-size: $numberkeyboard-key-finish-font-size;
          color: $numberkeyboard-key-finish-font-color;
          background-color: $numberkeyboard-key-finish-background-color;
        }

        .activeFinsh {
          font-size: $numberkeyboard-key-finish-font-size;
          color: $numberkeyboard-key-finish-font-color;
          background-color: $numberkeyboard-key-activeFinsh-background-color;
        }
      }
    }
  }
}
