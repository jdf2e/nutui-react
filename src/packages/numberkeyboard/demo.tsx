import React, { useState } from 'react'
import { useTranslate } from '@/sites/assets/locale'
import Cell from '../cell'
import Toast from '../toast'
import { NumberKeyboard } from './numberkeyboard'

const NumberKeyboardDemo = () => {
  const [translated] = useTranslate({
    'zh-CN': {
      basic: '默认键盘',
      sidebar: '带右侧栏键盘',
      confirmText: '支付',
      randomKeyOrder: '随机数键盘',
      title: '标题',
      withTitle: '带标题栏键盘',
      idNumberKeyboard: '身份证键盘',
    },
    'zh-TW': {
      basic: '默認鍵盤',
      sidebar: '帶右側欄鍵盤',
      confirmText: '支付',
      randomKeyOrder: '隨機數鍵盤',
      title: '標題',
      withTitle: '帶標題欄鍵盤',
      idNumberKeyboard: '身份證鍵盤',
    },
    'en-US': {
      basic: 'Default Keyboard',
      sidebar: 'Keyboard With Sidebar',
      confirmText: 'pay',
      randomKeyOrder: 'Random Key Order',
      title: 'title',
      withTitle: 'Show Keyboard With Title',
      idNumberKeyboard: 'Show IdNumber Keyboard',
    },
  })
  const [visible1, setVisible1] = useState(false)
  const [visible2, setVisible2] = useState(false)
  const [visible3, setVisible3] = useState(false)
  const [visible4, setVisible4] = useState(false)
  const [visible5, setVisible5] = useState(false)
  const onChange = (number: string) => {
    Toast.show(`输入：${number}`)
  }
  const onDelete = () => {
    Toast.show('删除')
  }
  return (
    <div className="demo">
      <h2>{translated.basic}</h2>
      <Cell
        title={translated.basic}
        onClick={() => {
          setVisible1(true)
        }}
      />
      <NumberKeyboard
        visible={visible1}
        onChange={onChange}
        onDelete={onDelete}
        onClose={() => setVisible1(false)}
      />
      <h2>{translated.sidebar}</h2>
      <Cell
        title={translated.sidebar}
        onClick={() => {
          setVisible2(true)
        }}
      />
      <NumberKeyboard
        visible={visible2}
        type="rightColumn"
        customKey={['.', 'x']}
        onChange={onChange}
        onDelete={onDelete}
        onClose={() => setVisible2(false)}
      />
      <h2>{translated.randomKeyOrder}</h2>
      <Cell
        title={translated.randomKeyOrder}
        onClick={() => {
          setVisible3(true)
        }}
      />
      <NumberKeyboard
        visible={visible3}
        randomKeys
        onChange={onChange}
        onDelete={onDelete}
        onClose={() => setVisible3(false)}
      />
      <h2>{translated.withTitle}</h2>
      <Cell
        title={translated.withTitle}
        onClick={() => {
          setVisible4(true)
        }}
      />
      <NumberKeyboard
        visible={visible4}
        title={translated.title}
        customKey={['.']}
        onChange={onChange}
        onDelete={onDelete}
        onClose={() => setVisible4(false)}
      />
      <h2>{translated.idNumberKeyboard}</h2>
      <Cell
        title={translated.idNumberKeyboard}
        onClick={() => {
          setVisible5(true)
        }}
      />
      <NumberKeyboard
        visible={visible5}
        customKey={['X']}
        onChange={onChange}
        onDelete={onDelete}
        onClose={() => setVisible5(false)}
      />
    </div>
  )
}

export default NumberKeyboardDemo
