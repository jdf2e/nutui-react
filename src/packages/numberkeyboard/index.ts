import { NumberKeyboard } from './numberkeyboard'

export default NumberKeyboard
