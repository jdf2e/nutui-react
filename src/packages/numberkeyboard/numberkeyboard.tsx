import React, { FunctionComponent, useMemo, useState } from 'react'
import classNames from 'classnames'
import { RectDown } from '@nutui/icons-react'
import Popup from '@/packages/popup'
import { PopupProps } from '@/packages/popup/popup'
import { useConfig } from '@/packages/configprovider'
import { ComponentDefaults } from '@/utils/typings'

export interface NumberKeyboardProps extends PopupProps {
  visible: boolean
  confirmText?: string
  type: 'default' | 'rightColumn'
  custom: Array<string>
  random: boolean
  onChange?: (value: string) => void
  onDelete?: () => void
  onClose: () => void
  onConfirm?: () => void
}
const defaultProps = {
  ...ComponentDefaults,
  visible: false,
  type: 'default',
  custom: [],
  random: false,
  onClose: () => {},
} as unknown as NumberKeyboardProps

export const NumberKeyboard: FunctionComponent<
  Partial<NumberKeyboardProps> &
    Omit<React.HTMLAttributes<HTMLDivElement>, 'onChange' | 'onClick' | 'title'>
> = (props) => {
  const { locale } = useConfig()
  const {
    title,
    confirmText,
    visible,
    type,
    custom,
    random,
    style,
    className,
    onChange,
    onDelete,
    onClose,
    onConfirm,
    ...rest
  } = props
  const classPrefix = 'nut-numberkeyboard'

  const getBasicKeys = () => {
    const keys = new Array(9).fill(0).map((_, index) => {
      return {
        id: String(index + 1),
        type: 'num',
      }
    })
    return random ? keys.sort(() => (Math.random() > 0.5 ? 1 : -1)) : keys
  }

  const getCustomKeys = () => {
    const customKeys = [
      { id: 'close', type: 'close' },
      { id: '0', type: 'num' },
      { id: 'delete', type: 'delete' },
    ]
    if (!custom) return customKeys
    if (custom.length > 0) {
      customKeys[0] = { id: custom[0], type: 'custom' }
    }
    if (custom.length > 1) {
      customKeys[2] = { id: custom[1], type: 'custom' }
    }
    return customKeys
  }

  const keysList = useMemo(() => {
    return [...getBasicKeys(), ...getCustomKeys()]
  }, [type, random, custom])

  const DeleteIcon = () => {
    return (
      <svg viewBox="0 0 1024 1024" width="36" height="36">
        <path
          d="M875.594 186.122H390.803a51.162 51.162 0 0 0-36.18 14.986L79.91 475.821a51.166 51.166 0 0 0 0 72.358l274.714 274.712a51.164 51.164 0 0 0 36.179 14.986h484.791c46.033 0 83.484-37.45 83.484-83.483V269.606c.001-46.033-37.45-83.484-83.483-83.484zm32.32 568.274c0 17.85-14.473 32.318-32.32 32.318H390.803L116.089 512l274.714-274.714h484.791c17.849 0 32.32 14.47 32.32 32.32v484.789z"
          fill="currentColor"
        />
        <path
          d="M753.945 360.214l-121.43 121.429-121.43-121.429s-16.062-8.224-30.356 6.072c-14.295 14.295-6.073 30.357-6.073 30.357l121.43 121.428L486.8 627.357s-8.222 16.062 6.072 30.357c14.297 14.296 30.358 6.072 30.358 6.072l109.286-109.285 109.286 109.285s16.062 8.224 30.357-6.072c14.295-14.295 6.07-30.357 6.07-30.357L668.944 518.072l121.431-121.43s8.22-16.061-6.074-30.356c-14.294-14.296-30.356-6.072-30.356-6.072z"
          fill="currentColor"
        />
      </svg>
    )
  }

  const NumberKeyboardKey = (props: { item: any }) => {
    const { item } = props
    const [active, setActive] = useState(false)
    const onTouchStart = () => {
      setActive(true)
    }
    const onTouchEnd = (item: { id: string; type: string }) => {
      setActive(false)
      if (item.type === 'num' || item.type === 'custom') {
        onChange && onChange(item.id)
      }
      if (item.type === 'close') {
        onClose && onClose()
      }
      if (item.type === 'delete') {
        onDelete && onDelete()
      }
      if (item.type === 'confirm') {
        onConfirm && onConfirm()
      }
    }
    return (
      <div key={item.id} className={`${classPrefix}__body__wrapper`}>
        <div
          className={classNames({
            key: true,
            active,
            close: item.type === 'close',
            delete: item.type === 'delete',
            confirm: item.type === 'confirm',
          })}
          onTouchStart={() => onTouchStart()}
          onTouchEnd={() => onTouchEnd(item)}
          onTouchCancel={() => onTouchEnd(item)}
        >
          {(item.type === 'num' || item.type === 'custom') && (
            <div>{item.id}</div>
          )}
          {item.type === 'delete' && <DeleteIcon />}
          {item.type === 'close' && <RectDown width={25} height={25} />}
          {item.type === 'confirm' && <>{confirmText || locale.done}</>}
        </div>
      </div>
    )
  }

  return (
    <Popup
      visible={visible}
      position="bottom"
      onOverlayClick={onClose}
      onCloseIconClick={onClose}
      overlayStyle={{ backgroundColor: 'rgba(0, 0, 0, 0)' }}
      {...rest}
    >
      <div className={classNames(classPrefix, className)} style={style}>
        {title && (
          <div className={`${classPrefix}__header`}>
            <h3 className={`${classPrefix}__header__title`}>{title}</h3>
            {type === 'default' && (
              <span
                className={`${classPrefix}__header__close`}
                onClick={onConfirm}
              >
                {locale.done}
              </span>
            )}
          </div>
        )}
        <div className={`${classPrefix}__body`}>
          <div className={`${classPrefix}__body__keys`}>
            {keysList?.map((item: any) => {
              return <NumberKeyboardKey key={item.id} item={item} />
            })}
          </div>
          {type === 'rightColumn' && (
            <div className={`${classPrefix}__sidebar`}>
              <NumberKeyboardKey
                key="delete"
                item={{
                  id: 'delete',
                  type: 'delete',
                }}
              />
              <NumberKeyboardKey
                key="confirm"
                item={{
                  id: 'confirm',
                  type: 'confirm',
                }}
              />
            </div>
          )}
        </div>
      </div>
    </Popup>
  )
}

NumberKeyboard.defaultProps = defaultProps
NumberKeyboard.displayName = 'NutNumberKeyboard'
