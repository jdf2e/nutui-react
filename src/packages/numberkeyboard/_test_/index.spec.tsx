import * as React from 'react'
import { act, render } from '@testing-library/react'
import { trigger } from '@/utils/test/event'
import '@testing-library/jest-dom'

import { NumberKeyboard } from '../numberkeyboard'

function clickKey(key: Parameters<typeof trigger>[0]) {
  return act(() => {
    trigger(key, 'touchstart')
    trigger(key, 'touchend')
  })
}
test('should match snapshot', async () => {
  const { container } = render(<NumberKeyboard visible />)
  expect(container).toMatchSnapshot()
})
test('should match custom snapshot', async () => {
  const { container } = render(
    <NumberKeyboard visible type="rightColumn" custom={['.', 'x', 'r']} />
  )
  expect(container.querySelector('.confirm')).toHaveTextContent('完成')
  expect(container).toMatchSnapshot()
})
test('default keyboard custom keyboard', async () => {
  const { container } = render(
    <NumberKeyboard visible custom={['.', 'x', 'r']} />
  )
  expect(container.querySelectorAll('.key')[10]).toHaveTextContent('0')
})
test('Has title and custom keyboard length greater than 1', async () => {
  const { container } = render(
    <NumberKeyboard visible title="标题" custom={['.', 'x', 'r']} />
  )
  expect(container.querySelectorAll('.key')[10]).toHaveTextContent('0')
})
test('Has title and custom keyboard length equal 1', async () => {
  const { container } = render(
    <NumberKeyboard visible title="标题" custom={['.']} />
  )
  expect(container.querySelectorAll('.key')[9]).toHaveTextContent('.')
})
test('should shuffle key order when using random-key prop', async () => {
  let value = 0
  const onChange = (v: string) => {
    value = Number(v)
  }
  const { container } = render(
    <NumberKeyboard visible random onChange={onChange} />
  )
  const keys: number[] = []
  const clickKeys: number[] = []

  for (let i = 0; i < 9; i++) {
    keys.push(i + 1)
    clickKey(container.querySelectorAll('.key')[i])
    clickKeys.push(value)
  }

  expect(keys.every((v, k) => keys[k] === clickKeys[k])).toEqual(false)
})
test('should emit delete event after clicking delete key', () => {
  const onDelete = jest.fn()
  const { container } = render(<NumberKeyboard visible onDelete={onDelete} />)
  clickKey(container.querySelectorAll('.key')[11])
  expect(onDelete).toBeTruthy()
})
test('should emit close event after clicking collapse key', () => {
  const onClose = jest.fn()
  const { container } = render(<NumberKeyboard visible onClose={onClose} />)
  clickKey(container.querySelectorAll('.key')[9])
  expect(onClose).toBeTruthy()
})
test('should render title and close button correctly', () => {
  const { container } = render(<NumberKeyboard visible title="默认键盘" />)
  const title = container.querySelector('.nut-numberkeyboard-header-title')
  expect(title).toHaveTextContent('默认键盘')
})

test('should render confirm button correctly', () => {
  const { container } = render(
    <NumberKeyboard visible type="rightColumn" confirmText="支付" />
  )
  const title = container.querySelector('.confirm')
  expect(title).toHaveTextContent('支付')
})
