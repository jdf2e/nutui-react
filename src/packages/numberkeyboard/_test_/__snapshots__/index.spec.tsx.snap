// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match customKey snapshot 1`] = `
<div>
  <div>
    <div
      class="overlay-fade-enter-active nut-overlay"
      style="background-color: rgba(0, 0, 0, 0); z-index: 2000; animation-duration: 0.3s;"
    />
    <div
      class="popup-bottom  undefined nut-popup"
      style="z-index: 2000; animation-duration: 0.3s;"
    >
      <div
        class="nut-numberkeyboard "
      >
        <div
          class="nut-numberkeyboard__body"
        >
          <div
            class="nut-numberkeyboard__body__keys"
          >
            <div
              class="key-board-wrapper"
            >
              <div
                class="key"
              >
                <div>
                  1
                </div>
              </div>
            </div>
            <div
              class="key-board-wrapper"
            >
              <div
                class="key"
              >
                <div>
                  2
                </div>
              </div>
            </div>
            <div
              class="key-board-wrapper"
            >
              <div
                class="key"
              >
                <div>
                  3
                </div>
              </div>
            </div>
            <div
              class="key-board-wrapper"
            >
              <div
                class="key"
              >
                <div>
                  4
                </div>
              </div>
            </div>
            <div
              class="key-board-wrapper"
            >
              <div
                class="key"
              >
                <div>
                  5
                </div>
              </div>
            </div>
            <div
              class="key-board-wrapper"
            >
              <div
                class="key"
              >
                <div>
                  6
                </div>
              </div>
            </div>
            <div
              class="key-board-wrapper"
            >
              <div
                class="key"
              >
                <div>
                  7
                </div>
              </div>
            </div>
            <div
              class="key-board-wrapper"
            >
              <div
                class="key"
              >
                <div>
                  8
                </div>
              </div>
            </div>
            <div
              class="key-board-wrapper"
            >
              <div
                class="key"
              >
                <div>
                  9
                </div>
              </div>
            </div>
            <div
              class="key-board-wrapper"
            >
              <div
                class="key"
              >
                <div>
                  .
                </div>
              </div>
            </div>
            <div
              class="key-board-wrapper"
            >
              <div
                class="key"
              >
                <div>
                  0
                </div>
              </div>
            </div>
            <div
              class="key-board-wrapper"
            >
              <div
                class="key"
              >
                <div>
                  x
                </div>
              </div>
            </div>
          </div>
          <div
            class="nut-numberkeyboard__sidebar"
          >
            <div
              class="key-board-wrapper"
            >
              <div
                class="key"
              >
                <img
                  alt=""
                  src="https://img11.360buyimg.com/imagetools/jfs/t1/129395/8/12735/2030/5f61ac37E70cab338/fb477dc11f46056c.png"
                />
              </div>
            </div>
            <div
              class="key-board-wrapper key-board-finish"
            >
              <div
                class="key finish"
              >
                完成
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`should match snapshot 1`] = `
<div>
  <div>
    <div
      class="overlay-fade-enter-active nut-overlay"
      style="background-color: rgba(0, 0, 0, 0); z-index: 2000; animation-duration: 0.3s;"
    />
    <div
      class="popup-bottom  undefined nut-popup"
      style="z-index: 2000; animation-duration: 0.3s;"
    >
      <div
        class="nut-numberkeyboard "
      >
        <div
          class="nut-numberkeyboard__body"
        >
          <div
            class="nut-numberkeyboard__body__keys"
          >
            <div
              class="key-board-wrapper"
            >
              <div
                class="key"
              >
                <div>
                  1
                </div>
              </div>
            </div>
            <div
              class="key-board-wrapper"
            >
              <div
                class="key"
              >
                <div>
                  2
                </div>
              </div>
            </div>
            <div
              class="key-board-wrapper"
            >
              <div
                class="key"
              >
                <div>
                  3
                </div>
              </div>
            </div>
            <div
              class="key-board-wrapper"
            >
              <div
                class="key"
              >
                <div>
                  4
                </div>
              </div>
            </div>
            <div
              class="key-board-wrapper"
            >
              <div
                class="key"
              >
                <div>
                  5
                </div>
              </div>
            </div>
            <div
              class="key-board-wrapper"
            >
              <div
                class="key"
              >
                <div>
                  6
                </div>
              </div>
            </div>
            <div
              class="key-board-wrapper"
            >
              <div
                class="key"
              >
                <div>
                  7
                </div>
              </div>
            </div>
            <div
              class="key-board-wrapper"
            >
              <div
                class="key"
              >
                <div>
                  8
                </div>
              </div>
            </div>
            <div
              class="key-board-wrapper"
            >
              <div
                class="key"
              >
                <div>
                  9
                </div>
              </div>
            </div>
            <div
              class="key-board-wrapper"
            >
              <div
                class="key lock"
              >
                <img
                  alt=""
                  src="https://img11.360buyimg.com/imagetools/jfs/t1/146371/38/8485/738/5f606425Eca239740/14f4b4f5f20d8a68.png"
                />
              </div>
            </div>
            <div
              class="key-board-wrapper"
            >
              <div
                class="key"
              >
                <div>
                  0
                </div>
              </div>
            </div>
            <div
              class="key-board-wrapper"
            >
              <div
                class="key delete"
              >
                <img
                  alt=""
                  src="https://img11.360buyimg.com/imagetools/jfs/t1/129395/8/12735/2030/5f61ac37E70cab338/fb477dc11f46056c.png"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
